### PR TITLE
Handle nested declarations

### LINKED ISSUE #34: Investigate inner classes
I have a feeling that we are not preserving "simpleNames" from ClassNames, so this could cause issues with inner classes

### PR DESCRIPTION
**Pull request by serras**
Fixes #34

I've get rid of `name`, and now at each point we carefully consider whether we need a `qfName` (fully qualified) or a `baseName` (just the final element).

Another change was to "flatten" the name of mutable copy classes, since we cannot add them to an already-existing class. So if for example we have:

```kotlin
class A { data class B(...) }
```

the name for the mutable copy class is `A$B$Mutable`.

### CODE REVIEW

File: utils/kotlin-poet/src/main/kotlin/fp/serrano/kopykat/KotlinPoetExtensions.kt [hunk 1]
```diff
@@ -36,11 +37,23 @@ public fun FileSpec.writeTo(codeGenerator: CodeGenerator) {
...
+    }
 
 public fun TypeName.asTransformLambda(): LambdaTypeName =
   LambdaTypeName.get(parameters = arrayOf(this), returnType = this)
 
 public fun TypeName.asReceiverConsumer(): LambdaTypeName =
   LambdaTypeName.get(receiver = this, returnType = UNIT)
+
+public fun ClassName.flattenWithSuffix(suffix: String): ClassName {
+  val mutableSimpleName = (simpleNames + suffix).joinToString(separator = "$")
```
Comment (pablisco) [inline utils/kotlin-poet/src/main/kotlin/fp/serrano/kopykat/KotlinPoetExtensions.kt:57]: Nice!

File: ksp/src/main/kotlin/fp/serrano/kopykat/MutableCopy.kt [hunk 2]
```diff
@@ -66,27 +65,36 @@ internal fun FileCompilerScope.addFreezeFunction() {
...
+      "is ${type.qfName} -> this.copy(${properties.joinToString { "${it.baseName} = ${it.baseName}" }})"
     })
   }
 }
 
 internal fun FileCompilerScope.addCopyFunction(block: FunSpec.Builder.() -> Unit) {
   addInlinedFunction(name = "copy", receives = target.parameterized, returns = target.parameterized, block = block)
 }
+
+internal fun FileCompilerScope.addDslMarkerClass() {
```
Comment (pablisco) [inline ksp/src/main/kotlin/fp/serrano/kopykat/MutableCopy.kt:95]:
```suggestion
private fun FileCompilerScope.addDslMarkerClass() {
```

File: ksp/src/main/kotlin/fp/serrano/kopykat/CopyMap.kt [hunk 3]
```diff
@@ -3,6 +3,7 @@
 package fp.serrano.kopykat
 
 import com.squareup.kotlinpoet.FileSpec
+import fp.serrano.kopykat.utils.*
```
Comment (pablisco) [inline ksp/src/main/kotlin/fp/serrano/kopykat/CopyMap.kt:6]: I removed these on my settings:  



I've created a ticket to add auto-formatting settings to the repo 🤞 #36

Saves us going back and forth 😅

**Review by pablisco — COMMENTED** [review]
Wow! This is a much simpler solution compared to what I had in mind :heart\_eyes: 

Thank you for sorting this out. :pray:

File: ksp/src/main/kotlin/fp/serrano/kopykat/utils/KspExtensions.kt [hunk 4]
```diff
@@ -2,12 +2,21 @@ package fp.serrano.kopykat.utils
...
 
-internal val KSDeclaration.name get() = simpleName.asString()
+internal val KSDeclaration.baseName get() = simpleName.asString()
+
+internal val KSDeclaration.qfName get() = qualifiedName?.asString() ?: simpleName.asString()
 
 internal val KSClassDeclaration.sealedTypes get() = getSealedSubclasses()
 
+internal fun KSDeclarationContainer.allNestedDeclarations(): Sequence<KSDeclaration> =
+internal fun KSDeclarationContainer.allNestedDeclarations(): Sequence<KSDeclaration> =
```
Comment (pablisco) [inline ksp/src/main/kotlin/fp/serrano/kopykat/utils/KspExtensions.kt:15]: Sorry @serras My suggestion destroyed the build :grimacing:

File: ksp/src/main/kotlin/fp/serrano/kopykat/utils/KspExtensions.kt [hunk 5]
```diff
@@ -2,12 +2,24 @@ package fp.serrano.kopykat.utils
 
 import com.google.devtools.ksp.symbol.KSClassDeclaration
 import com.google.devtools.ksp.symbol.KSDeclaration
+import com.google.devtools.ksp.symbol.KSDeclarationContainer
 import com.google.devtools.ksp.symbol.KSPropertyDeclaration
 
-internal val KSDeclaration.name get() = simpleName.asString()
+internal val KSDeclaration.baseName get() = simpleName.asString()
+
+internal val KSDeclaration.qfName get() = qualifiedName?.asString() ?: simpleName.asString()
```
[the remark this answers is not in the document]
Comment (serras) [reply ksp/src/main/kotlin/fp/serrano/kopykat/utils/KspExtensions.kt]: Renamed it to `fullName`

File: ksp/src/main/kotlin/fp/serrano/kopykat/utils/GeneratedMarker.kt [hunk 6]
```diff
@@ -8,12 +8,12 @@ import com.squareup.kotlinpoet.asTypeName
...
+    PropertySpec.builder(MARKER, UnitTypeName).addModifiers(PRIVATE).initializer("Unit").build()
   )
 }
 
 internal fun KSFile.hasGeneratedMarker(): Boolean =
-  declarations.filterIsInstance<KSPropertyDeclaration>().any { it.name == Marker }
+  declarations.filterIsInstance<KSPropertyDeclaration>().any { it.simpleName.asString() == MARKER }
 
-private const val Marker = "transformativeGenerated"
+private const val MARKER = "transformativeGenerated"
```
[the remark this answers is not in the document]
Comment (serras) [reply ksp/src/main/kotlin/fp/serrano/kopykat/utils/GeneratedMarker.kt]: I just changed it becuase Detekt was screaming at me, but I've undone it and told Detekt to let me

File: utils/kotlin-poet/src/main/kotlin/fp/serrano/kopykat/KotlinPoetExtensions.kt [hunk 7]
```diff
@@ -36,11 +37,23 @@ public fun FileSpec.writeTo(codeGenerator: CodeGenerator) {
...
+public fun ClassName.append(name: String): ClassName =
+  ClassName(packageName, simpleNames + name)
+
 public val KSDeclaration.className: ClassName
-  get() = ClassName(packageName = packageName.asString(), simpleName.asString())
+  get() =
+    when (parentDeclaration) {
+      is KSClassDeclaration -> parentDeclaration!!.className.append(simpleName.asString())
+      else -> ClassName(packageName = packageName.asString(), simpleName.asString())
+    }
```
[the remark this answers is not in the document]
Comment (serras) [reply utils/kotlin-poet/src/main/kotlin/fp/serrano/kopykat/KotlinPoetExtensions.kt]: I prefer the stable name, I've already had enough symbol soup in my Haskell days...

File: ksp/src/main/kotlin/fp/serrano/kopykat/utils/KspExtensions.kt (hunk 4 above)
Comment (serras) [reply ksp/src/main/kotlin/fp/serrano/kopykat/utils/KspExtensions.kt]: Curse You!

**Pull request merged**